### PR TITLE
extend load balancer context interface to provide override host

### PR DESCRIPTION
**Pull request by wbpcode**
Signed-off-by: wbpcode <wbphub@live.com>

Commit Message: extend load balancer context interface to provide override host
Additional Description:

Continuation of #17290. Part of #16698. In the #17290, we implement a cross priority host map for fast host searching and add it to all types of Cluster. This PR extends load balancer context interface to provider override host. Then the load balancer can select host based on this override host directly and bypass algorithm selection.

Get more detailed design of stateful session sticky by ref https://docs.google.com/document/d/1IU4b76AgOXijNa4sew1gfBfSiOMbZNiEt5Dhis8QpYg/edit?usp=sharing

Risk Level: Mid.
Testing: Add.
Docs Changes: N/A.
Release Notes: N/A.
Platform Specific Features: N/A.

### CODE REVIEW

File: envoy/upstream/load_balancer.h [hunk 1]
```diff
@@ -83,6 +83,16 @@ class LoadBalancerContext {
    * Returns the transport socket options which should be applied on upstream connections
    */
   virtual Network::TransportSocketOptionsConstSharedPtr upstreamTransportSocketOptions() const PURE;
+
+  using OverrideHostStatus = uint32_t;
```
Comment (mattklein123) [inline envoy/upstream/load_balancer.h:87]: This still needs documentation on how the bitfield works. It's not clear from reading this.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks, this LGTM with a small doc comment. Nice work. Please double check that you have test coverage on all of the new functionality across all of the different LB types. (You can check the coverage report in CI.)

@snowp can you take a pass also?

/wait

**Review by mattklein123 — APPROVED** [review]
Thanks LGTM pending @snowp review.

File: source/common/upstream/load_balancer_impl.cc [hunk 2]
```diff
@@ -507,8 +515,52 @@ bool ZoneAwareLoadBalancerBase::earlyExitNonLocalityRouting() {
...
+  if (!override_host.has_value()) {
+    return nullptr;
+  }
+
+  if (host_map == nullptr) {
+    return nullptr;
+  }
+
+  auto host_iter = host_map->find(override_host.value().first);
+  HostConstSharedPtr host = host_iter != host_map->end() ? host_iter->second : nullptr;
```
Comment (snowp) [inline source/common/upstream/load_balancer_impl.cc:547]: I think doing an early return here and breaking up L549 into two if checks would make it a bit easier to follow

File: source/common/upstream/load_balancer_impl.cc [hunk 3]
```diff
@@ -507,8 +515,52 @@ bool ZoneAwareLoadBalancerBase::earlyExitNonLocalityRouting() {
...
+
+  auto override_host = context->overrideHostToSelect();
+  if (!override_host.has_value()) {
+    return nullptr;
+  }
+
+  if (host_map == nullptr) {
+    return nullptr;
+  }
+
```
Comment (snowp) [inline source/common/upstream/load_balancer_impl.cc:545]: When would this be nullptr? Don't we always have a host_map?

File: source/common/upstream/load_balancer_impl.h [hunk 4]
```diff
@@ -40,24 +40,7 @@ class LoadBalancerBase : public LoadBalancer {
   choosePriority(uint64_t hash, const HealthyLoad& healthy_per_priority_load,
                  const DegradedLoad& degraded_per_priority_load);
 
-  HostConstSharedPtr chooseHost(LoadBalancerContext* context) override;
```
Comment (snowp) [inline source/common/upstream/load_balancer_impl.h:43]: Just so I understand, why did you move this down in the class hierarchy? Thanks

File: source/common/upstream/thread_aware_lb_impl.cc [hunk 5]
```diff
@@ -141,6 +144,12 @@ ThreadAwareLoadBalancerBase::LoadBalancerImpl::chooseHost(LoadBalancerContext* c
     return nullptr;
   }
 
+  HostConstSharedPtr host;
+  if (host = LoadBalancerContextBase::selectOverrideHost(cross_priority_host_map_.get(), context);
```
Comment (snowp) [inline source/common/upstream/thread_aware_lb_impl.cc:148]: I would either initialize host outside the if-block, or move the declaration inside. Right now I'm not seeing much benefit to this pattern

File: source/common/upstream/load_balancer_impl.h [hunk 6]
```diff
@@ -165,6 +149,11 @@ class LoadBalancerBase : public LoadBalancer {
 
 class LoadBalancerContextBase : public LoadBalancerContext {
 public:
+  static bool validateOverrideHostStatus(Host::Health health, OverrideHostStatus status);
+
+  static HostConstSharedPtr selectOverrideHost(const HostMap* host_map,
+                                               LoadBalancerContext* context);
```
Comment (snowp) [inline source/common/upstream/load_balancer_impl.h:155]: I wonder if it makes sense to have this be on the LoadBalancerBase, it just seems a tad bit odd to have a static function on a class that accepts a pointer to said class. Not a big deal if you prefer it here

File: test/common/upstream/load_balancer_impl_test.cc [hunk 7]
```diff
@@ -540,6 +568,99 @@ TEST_P(LoadBalancerBaseTest, BoundaryConditions) {
...
+    LoadBalancerContext::OverrideHost override_host{"1.2.3.4", HealthyStatus | DegradedStatus};
+    EXPECT_CALL(context, overrideHostToSelect())
+        .WillOnce(Return(absl::make_optional(override_host)));
+
+    // Mock membership update and update host map shared pointer in the lb.
+    auto host_map = std::make_shared<HostMap>();
+    priority_set_.cross_priority_host_map_ = host_map;
+    host_set_.runCallbacks({}, {});
+
+    // Expected host is not exist in the host map and then `chooseHostOnce` will be called.
```
Comment (snowp) [inline test/common/upstream/load_balancer_impl_test.cc:624]: I'd do this as `// The expected host does not exist in the host map, therefore ...`

File: test/common/upstream/load_balancer_impl_test.cc [hunk 8]
```diff
@@ -540,6 +568,99 @@ TEST_P(LoadBalancerBaseTest, BoundaryConditions) {
...
+    EXPECT_CALL(context, overrideHostToSelect())
+        .WillOnce(Return(absl::make_optional(override_host)));
+
+    // Mock membership update and update host map shared pointer in the lb.
+    auto host_map = std::make_shared<HostMap>();
+    host_map->insert({"1.2.3.4", mock_host});
+    priority_set_.cross_priority_host_map_ = host_map;
+    host_set_.runCallbacks({}, {});
+
+    // Host status does not match the expected host status and then `chooseHostOnce` will be called.
```
Comment (snowp) [inline test/common/upstream/load_balancer_impl_test.cc:642]: then -> therefore

File: test/common/upstream/load_balancer_impl_test.cc [hunk 9]
```diff
@@ -1904,6 +2025,135 @@ TEST(LoadBalancerSubsetInfoImplTest, KeysSubsetEqualKeysInvalid) {
                             "fallback_keys_subset cannot be equal to keys");
 }
 
+TEST(LoadBalancerContextBaseTest, LoadBalancerContextBaseTest) {
+  {
+    auto context = LoadBalancerContextBase();
```
Comment (snowp) [inline test/common/upstream/load_balancer_impl_test.cc:2030]: `LoadBalancerContextBase context;` would be more idiomatic here

File: test/common/upstream/load_balancer_impl_test.cc [hunk 10]
```diff
@@ -33,6 +34,34 @@ namespace Envoy {
 namespace Upstream {
 namespace {
 
+static constexpr uint32_t UnhealthyStatus = 1u << static_cast<size_t>(Host::Health::Unhealthy);
+static constexpr uint32_t DegradedStatus = 1u << static_cast<size_t>(Host::Health::Degraded);
+static constexpr uint32_t HealthyStatus = 1u << static_cast<size_t>(Host::Health::Healthy);
+
+LoadBalancerContext::OverrideHostStatus createOverrideHostStatus(
```
Comment (snowp) [inline test/common/upstream/load_balancer_impl_test.cc:41]: Maybe I'm missing something but this doesn't seem to be used?

**Review by snowp — CHANGES_REQUESTED** [review]
Thanks this looks pretty good to me, just a few comments

File: source/common/upstream/load_balancer_impl.cc (hunk 2 above)
Comment (wbpcode) [reply source/common/upstream/load_balancer_impl.cc:547]: get it.

File: source/common/upstream/load_balancer_impl.cc (hunk 3 above)
Comment (wbpcode) [reply source/common/upstream/load_balancer_impl.cc:545]: 1. Before the first priority update, it is nullptr.
2. The host maps of all sub-LBs of subset lb are nullptr. 

So I think keeping this check will make the code safer.

File: source/common/upstream/load_balancer_impl.h (hunk 4 above)
Comment (wbpcode) [reply source/common/upstream/load_balancer_impl.h:43]: This question is a bit more complicated.

Both `ThreadAwareLoadBalancerBase` and `ZoneAwareLoadBalancerBase` need a cross priority host map. But the use of the two is completely different. The former needs to ensure that the host map is thread-safe, while the latter does not need to consider this issue at all.

This led me to finally decide to add a member of cross_priority_host_map_ to `ThreadAwareLoadBalancerBase` and `ZoneAwareLoadBalancerBase` separately, instead of adding a member of cross_priority_host_map_ to the common base class `LoadBalancerBase` of the two.

But this brings a new problem, that is, the `chooseHost` method in `LoadBalancerBase` needs to access cross_priority_host_map_ to search for override host.
Although I know that the base class can access the members of the derived class through virtual functions, but there is no need to be so complicated.

The `chooseHost` method in `LoadBalancerBase` is only meaningful to `ZoneAwareLoadBalancerBase` and its derived classes, so there is no need to implement it in `LoadBalancerBase`.

So, I finally moved `chooseHost` to `ZoneAwareLoadBalancerBase`, so that it will not affect the function, but also can directly access cross_priority_host_map_ in `chooseHost`, and it is simpler and cleaner.

File: source/common/upstream/thread_aware_lb_impl.cc (hunk 5 above)
Comment (wbpcode) [reply source/common/upstream/thread_aware_lb_impl.cc:148]: get it.

File: test/common/upstream/load_balancer_impl_test.cc (hunk 7 above)
Comment (wbpcode) [reply test/common/upstream/load_balancer_impl_test.cc:624]: Thanks.

File: test/common/upstream/load_balancer_impl_test.cc (hunk 8 above)
Comment (wbpcode) [reply test/common/upstream/load_balancer_impl_test.cc:642]: Thanks.

File: test/common/upstream/load_balancer_impl_test.cc (hunk 9 above)
Comment (wbpcode) [reply test/common/upstream/load_balancer_impl_test.cc:2030]: get it.

File: test/common/upstream/load_balancer_impl_test.cc (hunk 10 above)
Comment (wbpcode) [reply test/common/upstream/load_balancer_impl_test.cc:41]: I will check it. Thanks.

Comment (wbpcode) [reply test/common/upstream/load_balancer_impl_test.cc:41]: It is used by some new added test case.

https://github.com/wbpcode/envoy/blob/0420d3012a97ee97948bfe824b3f2015725941ca/test/common/upstream/load_balancer_impl_test.cc#L2051-L2056

File: source/common/upstream/load_balancer_impl.h (hunk 6 above)
Comment (wbpcode) [reply source/common/upstream/load_balancer_impl.h:155]: I haven't found out a better position to place these util functions. I have no particular insistence on this, I think we can move it as needed in the future.

File: test/common/upstream/load_balancer_impl_test.cc (hunk 10 above)
Comment (snowp) [reply test/common/upstream/load_balancer_impl_test.cc:41]: Those tests seem to be testing `createOverrideHostStatus` itself, can we just remove it nothing is actually using it?

File: source/common/upstream/load_balancer_impl.h (hunk 4 above)
Comment (snowp) [reply source/common/upstream/load_balancer_impl.h:43]: Thanks for explaining! This makes sense to me

File: source/common/upstream/load_balancer_impl.cc (hunk 3 above)
Comment (snowp) [reply source/common/upstream/load_balancer_impl.cc:545]: Could we document this via a comment somewhere? Would be good to let people reading the code understand the need for a null host_map, thanks!

File: test/common/upstream/load_balancer_impl_test.cc (hunk 10 above)
Comment (wbpcode) [reply test/common/upstream/load_balancer_impl_test.cc:41]: Ha, you are right. :rofl:

**Review by snowp — APPROVED** [review]
LGTM, thanks!
